src/pages/VeSDL/index.tsx

```diff
@@ -4,52 +4,245 @@ import {
+  const [initialLoading, setInitialLoading] = useState<boolean>(false)
+
+  const { account, chainId, library } = useActiveWeb3React()
   const { t } = useTranslation()
-  const handleChange = (value: string) => {
-    setSdlTokenRawVal(value)
+
+  // Get votingEscrow contract from ABI
+  const votingEscrowContract = useMemo(() => {
+    if (chainId && library) {
+      return getContract(
+        VOTING_ESCROW_CONTRACT_ADDRESS?.[chainId],
+        VOTING_ESCROW_CONTRACT_ABI as ContractInterface,
+        library,
+        account ?? undefined,
+      ) as VotingEscrow
+    }
+  }, [account, chainId, library])
+
+  const sdlContract = useMemo(() => {
+    if (chainId && library) {
+      return getContract(
+        SDL_TOKEN_ADDRESSES[chainId],
+        SDL_TOKEN_ABI,
+        library,
+        account ?? undefined,
+      ) as Sdl
+    }
+  }, [account, chainId, library])
+
+  const fetchData = useCallback(async () => {
+    if (account) {
+      const sdlTokenBal = await sdlContract?.balanceOf(account)
+      setSDLToken((prev) => ({
+        ...prev,
+        sdlTokenInputVal: "",
+        maxBalance: formatUnits(sdlTokenBal || Zero),
+      }))
+      const vesdlBal = await votingEscrowContract?.["balanceOf(address)"](
+        account,
+      )
+      setVeSdlTokenVal(vesdlBal || Zero)
+
+      const prevLockEnd =
+        (await votingEscrowContract?.locked__end(account)) || Zero
+      setLockEnd(
+        !prevLockEnd.isZero() ? new Date(prevLockEnd.toNumber() * 1000) : null,
+      )
+
+      const veSdlToken = await votingEscrowContract?.locked(account)
+      setLockedSDLVal(veSdlToken?.amount || Zero)
+    }
+  }, [account, sdlContract, votingEscrowContract])
+
+  function calculateLockAmount() {
+    if (!unlockDate) return
+    if (isNaN(unlockDate.valueOf())) return
+    const currentTimestamp = getUnixTime(new Date())
+    const expireTimestamp = getUnixTime(unlockDate)
+    const totalAmount = sdlTokenValue.add(lockedSDLVal)
+    const roundedExpireTimestamp = BigNumber.from(expireTimestamp)
+      .div(WEEK)
+      .mul(WEEK)
+
+    const estLockValue = totalAmount
+      .mul(BigNumber.from(roundedExpireTimestamp).sub(currentTimestamp))
+      .div(MAXTIME)
+    return formatUnits(estLockValue)
+  }
+
+  const getPenalty = () => {
+    if (!lockEnd) return
+    const leftTime = getUnixTime(lockEnd) - getUnixTime(new Date())
+    const penaltyRatio = Math.min(
+      (MULTIPLIER * 3) / 4,
+      (MULTIPLIER * leftTime) / MAXTIME,
+    ).toFixed()
+
+    const penaltyAmount = lockedSDLVal
+      .mul(BigNumber.from(penaltyRatio))
+      .div(BigNumber.from(MULTIPLIER))
+    return formatUnits(penaltyAmount)
+  }
+
+  const handleLock = async () => {
+    if (!account || !chainId || !votingEscrowContract?.address) return
+    try {
+      const unlockTimeStamp = unlockDate ? getUnixTime(unlockDate) : null
+      if (sdlTokenValue.gt(Zero))
+        await sdlContract?.approve(votingEscrowContract?.address, sdlTokenValue)
```

Comment:
please use `checkAndApproveTokenForTrade` to avoid asking for approval when it's not needed